java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java

```diff
@@ -416,56 +441,14 @@ private boolean isArgumentMatchesFlagName(
    * 3. Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
    * 4. Qualified enum name with method call (e.g. "TestExperimentName.STALE_FLAG.getKey()")
    */
-  private boolean isMatchingEnumFieldValue(ExpressionTree argTree, VisitorState state) {
-    // Skip if no enum records are configured, to avoid excessive AST lookups
-    if (!this.config.hasEnumRecords()) {
-      return false;
-    }
-
-    Symbol sym;
-    switch (argTree.getKind()) {
-      case IDENTIFIER:
-        // Case 1: Unqualified enum constant (e.g. "STALE_FLAG")
-        sym = ASTHelpers.getSymbol(argTree);
-        break;
-      case MEMBER_SELECT:
-        // Case 2: Qualified enum name (e.g. "TestExperimentName.STALE_FLAG")
-        MemberSelectTree memberTree = (MemberSelectTree) argTree;
-        sym = ASTHelpers.getSymbol(memberTree);
-        break;
-      case METHOD_INVOCATION:
-        // Because the enum constant itself is being removed, we can remove any method called from
-        // that enum constant and not only specific, configured ones
-        MethodInvocationTree methodTree = (MethodInvocationTree) argTree;
-        if (methodTree.getMethodSelect().getKind() != Kind.MEMBER_SELECT) {
-          return false;
-        }
-        MemberSelectTree methodMemberTree = (MemberSelectTree) methodTree.getMethodSelect();
-        if (methodMemberTree.getExpression().getKind() == Kind.IDENTIFIER) {
-          // Case 3: Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
-          IdentifierTree identifier = (IdentifierTree) methodMemberTree.getExpression();
-          sym = ASTHelpers.getSymbol(identifier);
-        } else if (methodMemberTree.getExpression().getKind() == Kind.MEMBER_SELECT) {
-          // Case 4: Qualified enum name with method call (e.g.
-          // "TestExperimentName.STALE_FLAG.getKey()")
-          MemberSelectTree memberTreeExpression =
-              (MemberSelectTree) methodMemberTree.getExpression();
-          sym = ASTHelpers.getSymbol(memberTreeExpression);
-        } else {
-          return false;
-        }
-        break;
-      default:
-        return false;
-    }
 
-    if (sym.getKind() != ElementKind.ENUM_CONSTANT) {
-      return false;
-    }
-
-    Symbol.ClassSymbol enclosingClass = ASTHelpers.enclosingClass(sym);
-
-    return isEnumConstantMatchingFlagName(sym.getSimpleName().toString(), enclosingClass, state);
+  private Matcher<ExpressionTree> isMatchingEnumFieldValue(VisitorState state) {
+    Matcher<ExpressionTree> enumFieldValueMatcher =
+        symbolMatcher(
+            (sym, st) ->
+                isEnumConstantMatchingFlagName(
+                    sym.getSimpleName().toString(), ASTHelpers.enclosingClass(sym), state));
+    return anyOf(enumFieldValueMatcher, contains(ExpressionTree.class, enumFieldValueMatcher));
```

Comment:
Wouldn't this over-match any expression that includes the flag? That might be too aggressive, depending on where this is called (compared with the four cases of the original code).

For example, this matcher would match on:

```
TestExperimentName.STALE_FLAG || TestExperimentName.OTHER_FLAG
```

Would it not?

How about:

```
TestExperimentName.OTHER_FLAG.hasFoo(TestExperimentName.STALE_FLAG.getFoo())
```

?

Not 100% where this is being called from and if it's safe because of that, but in general this is matching a lot more aggressively than: enum constant alone or method call on enum constant. At the very least the documentation comments need to be moves around and adjusted (a more general javadoc for this method, the 4 cases above where they are validated to hold)

Reply:
I have addressed this. I have made the matcher as "narrow" as it was before the refactoring.

Reply:
As per our discussion online, you can observe that I have made the matchers "not lazy" too.